### DISCUSSION TITLE
can swig generate c++ wrapper of c code

### DISCUSSION THREAD

**Discussion by gogo2464** [swig/swig#2839] [Q&A] [answered]
Hi. There is a work on generating C wrapper from C++. Could I also generate c++ bindings from C code in swig please?

**Comment by FawwazRaza** [comment]
Yes, you can use SWIG (Simplified Wrapper and Interface Generator) to generate C++ bindings from C code. SWIG supports generating wrappers for C and C++ code interchangeably, allowing you to create interfaces in either language.

**Reply by gogo2464** [reply]
thanks but how. What is the command to generate c++ then? does not seems to be -c++

**Reply by gogo2464** [reply]
@Fawwaz-rz do I use -c++ command?

**Comment by ojwb** [comment, marked as answer]
If you're wanting SWIG to generate `foo_wrap.cxx` instead of `foo_wrap.c`  while generating code one of the supported target languages (Perl, PHP, Python, etc) then I'd recommend just using `-c++` even if your input is actually C code.  There are a few corner cases where SWIG won't parse valid C as C++ (e.g. `int class();`) but it's unlikely you'll hit these, especially if wrapping the public headers of a third party C library as generally these will be written to be usable from C++ too.

There is a `-c++out` option, but I would avoid that if you possibly can as it is completely undocumented and has no direct test coverage.  It hooks into the internal mechanism used by target languages which require C++ output, and there are known issues with that (#2818).

If you're asking about C++ as a target language (i.e. wrapping a C API to make a C++ one) then SWIG does not support this.  The idea has been raised before and was on the list of project ideas back when SWIG was taking part in GSoC, but I don't think anyone ever worked on it.  It's arguably less useful than most target languages as C++ natively supports calling a C API and has all the C types natively available, which isn't true for most of SWIG's target languages - the utility of a C++ target language would be that you could use SWIG's features to wrap the C API in a more object-oriented style.

**Reply by gogo2464** [reply]
I am on the very beginning of my project. Do you recommend me to rewrite everything in in C++ and then use C wrapper on the current PR branch?

My project is very ambitious and I have no experience with swig. The philosophy must follow.

**Reply by ojwb** [reply]
You've already asked that and I answered it.  I'll repeat my answer:

> A C API + %extend is the only one of your two approaches that's currently supported by released versions of SWIG, so that seems the better option until this branch gets merged, which sadly may never happen.

I'm sorry if that wasn't the answer you wanted, but it's the answer that reflects current reality and repeatedly asking the question won't change that reality.  Only the C backend work actually getting merged would, and distracting SWIG maintainers by asking the same question over and over uses up our limited time so if anything makes that slightly less likely to happen.
